src/main/java/org/sagebionetworks/bridge/spring/util/EtagComponent.java

```diff
@@ -0,0 +1,167 @@
+        // or when a dependent object is deleted).
+        String etag = calculateEtag(context, session);
+        
+        if (requestEtag != null) {
+            if (requestEtag.equals(etag)) {
+                if (LOG.isDebugEnabled()) {
+                    LOG.debug("Returning 304 for etag: " + etag);
+                }
+                response.addHeader(HttpHeaders.ETAG, etag);
+                response.setStatus(304);
+                return null;
+            }
+        }
+        Object retValue = joinPoint.proceed();
+        if (etag != null) {
+            response.addHeader(HttpHeaders.ETAG, etag);
+            if (LOG.isDebugEnabled()) {
+                LOG.debug("Returning etag to response: " + etag);
+            }
+        }
+        return retValue;
+    }
+
+    private String calculateEtag(EtagContext context, UserSession session) {
+        // Collect timestamps for all the dependencies that determine freshness of etag
+        List<DateTime> timestamps = new ArrayList<>();
+        for (EtagCacheKey timestampKey : context.getCacheKeys()) {
+            int len = timestampKey.keys().length;
+            String[] resolvedKeyValues = new String[len];
+            for (int i=0; i < len; i++) {
+                // given getValue's behavior, this value will not be null;
+                resolvedKeyValues[i] = getValue(context, session, timestampKey.keys()[i]);
+            }
+            // Retrieve the timestamp under this key
+            CacheKey cacheKey = CacheKey.etag(timestampKey.model(), resolvedKeyValues);
+            DateTime timestamp = cacheProvider.getObject(cacheKey, DateTime.class);
+            if (timestamp == null) {
+                return null; // this is a cache miss, any miss means there is no etag
+            }
+            timestamps.add(timestamp.withZone(DateTimeZone.UTC));
+        }
+        String base = BridgeUtils.SPACE_JOINER.join(timestamps);
+        byte[] md5 = md5DigestUtils.digest(base.getBytes());
+        return Hex.encodeHexString(md5);
+    }
+    
+    private String getValue(EtagContext context, UserSession session, String fieldName) {
+        if (context.getArgValues().containsKey(fieldName)) {
+            String value = (String) context.getArgValues().get(fieldName);
+            if (value == null) {
+                throw new IllegalArgumentException(NO_VALUE_ERROR + fieldName);
+            }
+            return value;
+        }
+        String value = null;
+        if (APP_ID_FIELD.equals(fieldName)) {
+            value = session.getAppId();
+        }
+        if (USER_ID_FIELD.equals(fieldName)) {
+            value = session.getId();
+        }
+        if (ORG_ID_FIELD.equals(fieldName)) {
+            value = session.getParticipant().getOrgMembership();
+        }
+        if (value == null) {
+            throw new IllegalArgumentException(NO_VALUE_ERROR + fieldName);
+        }
+        return value;
```

Comment:
Would this be cleaner as a switch statement? Alternatively, a series of if-else if statements instead of individual ifs?